File: python/observability/observability.py
"""
Starter observability module for Python 3.11+ services.

Vendored from org template; promote to a PyPI package once 3+ repos
have used this unchanged for a release cycle.

Wires up:
  - structlog stdout JSON logging with standard fields
  - prometheus_client metrics at /metrics
  - OpenTelemetry tracing (OTLP exporter) with standard resource attrs

Pip packages required:
    structlog
    prometheus-client
    opentelemetry-sdk
    opentelemetry-exporter-otlp
    opentelemetry-instrumentation-fastapi   # or your framework
    opentelemetry-instrumentation-requests

Usage (FastAPI example):

    from fastapi import FastAPI
    from observability import init, ObservabilityConfig, mount_endpoints

    cfg = ObservabilityConfig(
        service_name="orders-service",
        service_version="1.4.2",
        environment="dev",
        otlp_endpoint="http://otel-collector:4317",
    )
    init(cfg)

    app = FastAPI()
    mount_endpoints(app, cfg, readiness_check=lambda: True)
"""

from __future__ import annotations

import logging
import os
import time
from dataclasses import dataclass
from typing import Callable, Optional

import structlog
from opentelemetry import trace
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.sdk.resources import Resource
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from prometheus_client import CONTENT_TYPE_LATEST, generate_latest

_START_TIME = time.time()


@dataclass(frozen=True)
class ObservabilityConfig:
    service_name: str
    service_version: str
    environment: str = "dev"
    otlp_endpoint: Optional[str] = None
    log_level: str = "INFO"


def init(cfg: ObservabilityConfig) -> None:
    """Configure structlog + OTel. Call once at process start."""
    _init_logging(cfg)
    _init_tracing(cfg)


def _init_logging(cfg: ObservabilityConfig) -> None:
    level = getattr(logging, cfg.log_level.upper(), logging.INFO)
    logging.basicConfig(format="%(message)s", level=level)

    def _add_standard_fields(_, __, event_dict):
        event_dict.setdefault("service", cfg.service_name)
        event_dict.setdefault("version", cfg.service_version)
        event_dict.setdefault("env", cfg.environment)

        span = trace.get_current_span()
        ctx = span.get_span_context() if span else None
        if ctx and ctx.is_valid:
            event_dict.setdefault("trace_id", format(ctx.trace_id, "032x"))
            event_dict.setdefault("span_id", format(ctx.span_id, "016x"))
        return event_dict

    structlog.configure(
        processors=[
            structlog.contextvars.merge_contextvars,
            structlog.processors.add_log_level,
            structlog.processors.TimeStamper(fmt="iso", key="ts"),
            _add_standard_fields,
            structlog.processors.EventRenamer("msg"),
            structlog.processors.JSONRenderer(),
        ],
        wrapper_class=structlog.make_filtering_bound_logger(level),
        logger_factory=structlog.PrintLoggerFactory(),
        cache_logger_on_first_use=True,
    )


def _init_tracing(cfg: ObservabilityConfig) -> None:
    resource = Resource.create({
        "service.name": cfg.service_name,
        "service.version": cfg.service_version,
        "deployment.environment": cfg.environment,
    })
    provider = TracerProvider(resource=resource)

    endpoint = cfg.otlp_endpoint or os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
    if endpoint:
        provider.add_span_processor(
            BatchSpanProcessor(OTLPSpanExporter(endpoint=endpoint, insecure=True))
        )

    trace.set_tracer_provider(provider)


def mount_endpoints(
    app,
    cfg: ObservabilityConfig,
    readiness_check: Optional[Callable[[], bool]] = None,
) -> None:
    """
    Mount /healthz, /readyz, /version, /metrics on a FastAPI/Starlette app.
    For other frameworks, copy the handlers and adapt.
    """
    from fastapi import Response  # local import; framework is optional

    @app.get("/healthz")
    def healthz():
        return {
            "status": "ok",
            "service": cfg.service_name,
            "version": cfg.service_version,
            "uptime_s": int(time.time() - _START_TIME),
        }

    @app.get("/version")
    def version():
        return {
            "service": cfg.service_name,
            "version": cfg.service_version,
            "env": cfg.environment,
        }

    @app.get("/readyz")
    def readyz():
        ok = readiness_check() if readiness_check else True
        return Response(
            content='{"status":"ok"}' if ok else '{"status":"down"}',
            status_code=200 if ok else 503,
            media_type="application/json",
        )

    @app.get("/metrics")
    def metrics():
        return Response(content=generate_latest(), media_type=CONTENT_TYPE_LATEST)

File: dotnet/observability/Observability.cs
// Observability.cs
// Starter observability module for .NET 8+ services.
// Vendored from org template; promote to a NuGet package once 3+ repos
// have used this unchanged for a release cycle.
//
// Wires up:
//   - Serilog stdout JSON logging with standard fields
//   - prometheus-net metrics at /metrics
//   - OpenTelemetry tracing (OTLP exporter) with standard resource attrs
//
// NuGet packages required:
//   Serilog.AspNetCore
//   Serilog.Formatting.Compact
//   Serilog.Enrichers.Environment
//   prometheus-net.AspNetCore
//   OpenTelemetry.Extensions.Hosting
//   OpenTelemetry.Instrumentation.AspNetCore
//   OpenTelemetry.Instrumentation.Http
//   OpenTelemetry.Exporter.OpenTelemetryProtocol

using Microsoft.AspNetCore.Builder;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.DependencyInjection;
using OpenTelemetry.Resources;
using OpenTelemetry.Trace;
using Prometheus;
using Serilog;
using Serilog.Formatting.Compact;

namespace Org.Observability;

public sealed class ObservabilityOptions
{
    public required string ServiceName { get; init; }
    public required string ServiceVersion { get; init; }
    public string Environment { get; init; } = "dev";
    public string? OtlpEndpoint { get; init; } // e.g. http://otel-collector:4317
}

public static class ObservabilityExtensions
{
    /// <summary>
    /// Registers logging, metrics, and tracing with the org standard fields.
    /// Call before <c>builder.Build()</c>.
    /// </summary>
    public static WebApplicationBuilder AddOrgObservability(
        this WebApplicationBuilder builder,
        ObservabilityOptions opts)
    {
        // --- Logging (Serilog, stdout JSON, standard fields) ---
        builder.Host.UseSerilog((ctx, lc) =>
        {
            lc.MinimumLevel.Information()
              .Enrich.FromLogContext()
              .Enrich.WithProperty("service", opts.ServiceName)
              .Enrich.WithProperty("version", opts.ServiceVersion)
              .Enrich.WithProperty("env", opts.Environment)
              .WriteTo.Console(new CompactJsonFormatter());
        });

        // --- Tracing (OTel, OTLP) ---
        builder.Services.AddOpenTelemetry()
            .ConfigureResource(r => r
                .AddService(opts.ServiceName, serviceVersion: opts.ServiceVersion)
                .AddAttributes(new[]
                {
                    new KeyValuePair<string, object>("deployment.environment", opts.Environment),
                }))
            .WithTracing(t =>
            {
                t.AddAspNetCoreInstrumentation()
                 .AddHttpClientInstrumentation();

                if (!string.IsNullOrWhiteSpace(opts.OtlpEndpoint))
                {
                    t.AddOtlpExporter(o => o.Endpoint = new Uri(opts.OtlpEndpoint));
                }
            });

        return builder;
    }

    /// <summary>
    /// Registers /metrics, /healthz, /readyz, /version endpoints.
    /// Call after <c>app.Build()</c>.
    /// </summary>
    public static WebApplication UseOrgObservability(
        this WebApplication app,
        ObservabilityOptions opts,
        Func<CancellationToken, Task<bool>>? readinessCheck = null)
    {
        app.UseSerilogRequestLogging();
        app.MapMetrics(); // /metrics

        app.MapGet("/healthz", () => Results.Ok(new
        {
            status = "ok",
            service = opts.ServiceName,
            version = opts.ServiceVersion,
        }));

        app.MapGet("/version", () => Results.Ok(new
        {
            service = opts.ServiceName,
            version = opts.ServiceVersion,
            env = opts.Environment,
        }));

        app.MapGet("/readyz", async (CancellationToken ct) =>
        {
            var ok = readinessCheck is null || await readinessCheck(ct);
            return ok
                ? Results.Ok(new { status = "ok" })
                : Results.StatusCode(503);
        });

        return app;
    }
}

// Usage in Program.cs:
//
// var builder = WebApplication.CreateBuilder(args);
// builder.AddOrgObservability(new ObservabilityOptions
// {
//     ServiceName = "orders-service",
//     ServiceVersion = ThisAssembly.Git.SemVer.Major + "." +
//                      ThisAssembly.Git.SemVer.Minor + "." +
//                      ThisAssembly.Git.SemVer.Patch,
//     Environment = builder.Environment.EnvironmentName,
//     OtlpEndpoint = builder.Configuration["OTEL_EXPORTER_OTLP_ENDPOINT"],
// });
//
// var app = builder.Build();
// app.UseOrgObservability(opts, readinessCheck: async ct =>
// {
//     // ping db, cache, etc.
//     return true;
// });
// app.Run();
